### PR TITLE
Adding tabSemanticsLabel to CupertinoLocalizations

### PR DESCRIPTION
**Pull request by Piinks**
## Description

Closing out this TODO: https://github.com/flutter/flutter/blob/418007d17221ea632efcbc11b0da7e9fd7164122/packages/flutter/lib/src/cupertino/bottom_tab_bar.dart#L216

This localizes the hint for the tab index to `CupertinoTabBar`.

## Related Issues

#13452

## Tests

Added unit test for `CupertinoLocalizations.tabSemanticsLabel`.
Updates affected tests.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

**WIP**

- [x] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [x] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by Piinks** [comment]
@shihaohong does this look about right? I'm going to add some tests, just wanted to check that I wired this up right. :)

**Comment by Piinks** [comment]
The script made a very large change to packages/flutter_localizations/lib/src/l10n/cupertino_kn.arb

File: packages/flutter_localizations/lib/src/l10n/cupertino_kn.arb [hunk 1]
```diff
@@ -1,22 +1,23 @@
...
-  "timerPickerHourLabelOther": "\u0c97\u0c82\u0c9f\u0cc6\u0c97\u0cb3\u0cc1",
-  "timerPickerMinuteLabelOne": "\u0ca8\u0cbf\u0cae\u0cbf\u002e",
-  "timerPickerMinuteLabelOther": "\u0ca8\u0cbf\u0cae\u0cbf\u002e",
-  "timerPickerSecondLabelOne": "\u0cb8\u0cc6\u002e",
-  "timerPickerSecondLabelOther": "\u0cb8\u0cc6\u002e",
-  "cutButtonLabel": "\u0c95\u0ca4\u0ccd\u0ca4\u0cb0\u0cbf\u0cb8\u0cbf",
-  "copyButtonLabel": "\u0ca8\u0c95\u0cb2\u0cbf\u0cb8\u0cbf",
-  "pasteButtonLabel": "\u0c85\u0c82\u0c9f\u0cbf\u0cb8\u0cbf",
-  "selectAllButtonLabel": "\u0c8e\u0cb2\u0ccd\u0cb2\u0cb5\u0ca8\u0ccd\u0ca8\u0cc2\u0020\u0c86\u0caf\u0ccd\u0c95\u0cc6\u0cae\u0cbe\u0ca1\u0cbf"
+  "datePickerHourSemanticsLabelOne": "$hour ಗಂಟೆ",
```
Comment (shihaohong) [inline packages/flutter_localizations/lib/src/l10n/cupertino_kn.arb:2]: This requires running `flutter/dev/tools/localization/bin/encode_kn_arb_files.dart`. There's a bug with emacs that causes the editor to crash when it detects text in Kannada script. 

I need to still make this part of the `gen_localizations` and `gen_date_localizations` scripts.

**Review by shihaohong — COMMENTED** [review]
Everything looks good! The only thing that's missing is running the following commands from your Flutter repo:
```
dart dev/tools/localization/bin/gen_localizations.dart --overwrite
dart dev/tools/localization/bin/encode_kn_arb_files.dart
```
The first command generates the localization delegates for each locale so that it will display 'TBD' for non-English locales for the tab label until translations are completed

The second command encodes Kannada script into Unicode escapes so that some versions of emacs do not crash when opening the Flutter repository.

File: packages/flutter_localizations/lib/src/l10n/cupertino_kn.arb (hunk 1 above)
Comment (Piinks) [reply packages/flutter_localizations/lib/src/l10n/cupertino_kn.arb:2]: Awesome! Thank you! Updates on the way.

**Comment by shihaohong** [comment]
For a unit test, add something similar to https://github.com/flutter/flutter/blob/master/packages/flutter_localizations/test/cupertino/translations_test.dart#L87 and I think you should be good!

**Comment by Piinks** [comment]
It looks like this may be a breaking change? I think I will have to update the existing tests with `CupertinoTabBar` to add `Localizations`. Unless these tests should be getting a default value (like the hard-coded value that was there)

**Comment by shihaohong** [comment]
I think you're right... I just realized that this also changes the behavior for the tab label in other locales (for example, using the `es` would now cause the text to show `TBD` instead of what we currently have).

**Comment by Piinks** [comment]
@Hixie and @HansMuller can you advise on this change?

This is breaking in that it
 - Changes the current hint provided in the `CupertinoTabBar` for various locales, and
 - Requires that the use of the `CupertinoTabBar` will require a `Localizations` parent to provide the appropriate hint.

@shihaohong mentioned that `CupertinoApp` should introduce the localizations delegate by default, which would make it less breaking in those cases... although the `CupertinoTabBar` may not be used in that context alone, as the framework tests prove out.

I'm still diagnosing a `Semantics` break here as well, just curious in general how this might be made less breaking. A similar change was made for `MaterialLocalizations` a few years back here: https://github.com/flutter/flutter/pull/13714

File: packages/flutter/lib/src/cupertino/localizations.dart [hunk 2]
```diff
@@ -150,6 +150,14 @@ abstract class CupertinoLocalizations {
   // The global version uses the translated string from the arb file.
   String get alertDialogLabel;
 
+  /// The accessibility label used on a tab in a [CupertinoTabBar].
+  ///
+  /// This message describes the index of the selected tab and how many tabs
+  /// there are, e.g. 'tab, 1 of 2' in United States English.
+  ///
+  /// `tabIndex` and `tabCount` must be greater than or equal to one.
+  String tabLabel({int tabIndex, int tabCount});
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/localizations.dart:159]: maybe semanticsTabLabel? or tabSemanticsLabel?

**Comment by Hixie** [comment]
CupertinoApp should definitely introduce the l10ns, just like MaterialApp does.

**Comment by Hixie** [comment]
i would check with devrel that they agree that having the Cupertino library follow the Material library's lead in how to do l10n is a good idea. assuming that they do, then a migration guide should be simple enough to provide.

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 3]
```diff
@@ -178,10 +179,13 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
             style: CupertinoTheme.of(context).textTheme.tabLabelTextStyle.copyWith(color: inactive),
             child: Padding(
               padding: EdgeInsets.only(bottom: bottomPadding),
-              child: Row(
-                // Align bottom since we want the labels to be aligned.
-                crossAxisAlignment: CrossAxisAlignment.end,
-                children: _buildTabItems(context),
+              child: Semantics(
+                explicitChildNodes: true,
+                child:Row(
```
Comment (shihaohong) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart:184]: nit:
```suggestion
                child: Row(
```

File: packages/flutter/test/cupertino/tab_scaffold_test.dart [hunk 4]
```diff
@@ -435,15 +435,22 @@ void main() {
...
+      child: Directionality(
+        textDirection: TextDirection.ltr,
+        child: CupertinoTabScaffold(
+          resizeToAvoidBottomInset: false,
+          tabBar: _buildTabBar(),
+          tabBuilder: (BuildContext context, int index) {
+            return const Placeholder();
+          },
+        ),
+      )
```
Comment (shihaohong) [inline packages/flutter/test/cupertino/tab_scaffold_test.dart:453]: nit: trailing comma
```suggestion
      ),
```

**Review by shihaohong — APPROVED** [review]
LGTM!

**Comment by Piinks** [comment]
Thanks for the guidance and review @shihaohong! 🎉 
I'm putting together a migration guide and gathering some feedback on the break. I'll mark this WIP in the meantime.

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 5]
```diff
@@ -204,6 +208,14 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
 
   List<Widget> _buildTabItems(BuildContext context) {
     final List<Widget> result = <Widget>[];
+    final CupertinoLocalizations localizations = CupertinoLocalizations.of(context);
+    assert(
+      localizations != null,
+      'CupertinoTabBar requires a Localizations parent in order to provide an'
+        'appropriate Semantics hint for tab indexing. A CupertinoApp will'
+        'provide the DefaultCupertinoLocalizations, or you can instantiate your'
+        'own Localizations.'
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart:217]:
```suggestion
      'CupertinoTabBar requires a Localizations parent in order to provide an'
        'appropriate Semantics hint for tab indexing. A CupertinoApp will '
        'provide the DefaultCupertinoLocalizations, or you can instantiate your '
        'own Localizations.'
```

File: packages/flutter/lib/src/cupertino/bottom_tab_bar.dart [hunk 6]
```diff
@@ -204,6 +208,14 @@ class CupertinoTabBar extends StatelessWidget implements PreferredSizeWidget {
 
   List<Widget> _buildTabItems(BuildContext context) {
     final List<Widget> result = <Widget>[];
+    final CupertinoLocalizations localizations = CupertinoLocalizations.of(context);
+    assert(
+      localizations != null,
+      'CupertinoTabBar requires a Localizations parent in order to provide an'
```
Comment (Piinks) [inline packages/flutter/lib/src/cupertino/bottom_tab_bar.dart:214]:
```suggestion
      'CupertinoTabBar requires a Localizations parent in order to provide an '
```

File: packages/flutter/test/cupertino/bottom_tab_bar_test.dart [hunk 7]
```diff
@@ -11,9 +11,16 @@ import '../widgets/semantics_tester.dart';
...
+        DefaultWidgetsLocalizations.delegate,
+        DefaultCupertinoLocalizations.delegate,
+      ],
+      child: Directionality(
+        textDirection: TextDirection.ltr,
+        child: widget,
+      ),
     ),
   );
 }
```
Comment (xster) [inline packages/flutter/test/cupertino/bottom_tab_bar_test.dart:26]: can you add one test for the semantics content of the tabs?

**Review by xster — COMMENTED** [review]
Great. Thanks Kate!

**Comment by xster** [comment]
@Hixie CupertinoApp does introduce localization

File: packages/flutter/test/cupertino/bottom_tab_bar_test.dart (hunk 7 above)
Comment (Piinks) [reply packages/flutter/test/cupertino/bottom_tab_bar_test.dart:26]: There is already a test for that which was checking the value of the hard-coded string, now it is checking the localized result. :) 
https://github.com/flutter/flutter/blob/0549ab239009e6f76f95881364e935cd345a4da7/packages/flutter/test/cupertino/bottom_tab_bar_test.dart#L360

**Pull request merged**